crates/rattler_installs_packages/src/index/package_database.rs

```diff
@@ -111,6 +127,390 @@ impl PackageDb {
+
+        let dummy_sdist_file_name = SDistFilename {
+            distribution,
+            version,
+            format,
+        };
+
+        let dummy_sdist = SDist::new(dummy_sdist_file_name, Box::new(bytes))?;
+
+        let wheel_metadata = wheel_builder
+            .get_sdist_metadata(&dummy_sdist)
+            .await
+            .into_diagnostic()?;
+
+        // construct a real sdist filename
+        let sdist_filename = SDistFilename {
+            distribution: wheel_metadata.1.name.clone(),
+            version: wheel_metadata.1.version.clone(),
+            format,
+        };
+        let filename = ArtifactName::SDist(sdist_filename.clone());
+
+        Ok((wheel_metadata, filename))
+    }
+
+    /// Get artifact by file URL
+    pub async fn get_http_artifact<'a, 'i, P: Into<NormalizedPackageName>>(
+        &self,
+        p: P,
+        url: Url,
+        wheel_builder: &WheelBuilder<'a, 'i>,
+    ) -> miette::Result<&IndexMap<PypiVersion, Vec<ArtifactInfo>>> {
+        let str_name = url.path();
+        let url_hash = url.fragment().and_then(parse_hash);
+
+        let normalized_package_name = p.into();
+
+        // Get the contents of the artifact
+        let artifact_bytes = self
+            .http
+            .request(
+                url.clone(),
+                Method::GET,
+                HeaderMap::default(),
+                CacheMode::Default,
+            )
+            .await?;
+
+        let mut bytes = artifact_bytes
+            .into_body()
+            .into_local()
+            .await
+            .into_diagnostic()?;
+
+        let artifact_hash = {
+            let mut bytes_for_hash = vec![];
+            bytes.rewind().into_diagnostic()?;
+            bytes.read_to_end(&mut bytes_for_hash).into_diagnostic()?;
+            bytes.rewind().into_diagnostic()?;
+            ArtifactHashes {
+                sha256: Some(rattler_digest::compute_bytes_digest::<Sha256>(
+                    bytes_for_hash,
+                )),
+            }
+        };
+
+        if let Some(hash) = url_hash.clone() {
+            assert_eq!(hash, artifact_hash);
+        };
+
+        let (filename, data_bytes, metadata) = match str_name.ends_with(".whl") {
+            true => {
+                let wheel = Wheel::from_url_and_bytes(url.path(), &normalized_package_name, bytes)?;
+
+                let filename = ArtifactName::Wheel(wheel.name().clone());
+                let (data_bytes, metadata) = wheel.metadata()?;
+
+                (filename, data_bytes, metadata)
+            }
+            false => {
+                let (wheel_metadata, filename) = self
+                    .get_sdist_from_http(
+                        &normalized_package_name,
+                        url.clone(),
+                        bytes,
+                        wheel_builder,
+                    )
+                    .await?;
+
+                (filename, wheel_metadata.0, wheel_metadata.1)
+            }
+        };
```

Comment:
You can use an if statment that returns explicitly, thats somewhat more ergonomic e.g:
```suggestion
        let (filename, data_bytes, metadata) = if str_name.ends_with(".whl") {
                let wheel = Wheel::from_url_and_bytes(url.path(), &normalized_package_name, bytes)?;

                let filename = ArtifactName::Wheel(wheel.name().clone());
                let (data_bytes, metadata) = wheel.metadata()?;

                (filename, data_bytes, metadata)
            }
            else 
            {
                let (wheel_metadata, filename) = self
                    .get_sdist_from_http(
                        &normalized_package_name,
                        url.clone(),
                        bytes,
                        wheel_builder,
                    )
                    .await?;

                (filename, wheel_metadata.0, wheel_metadata.1)
            }
        };